File: adapter/src/main/java/com/ashlikun/adapter/recyclerview/multiltem/MultipleSingAdapter.java
package com.ashlikun.adapter.recyclerview.multiltem;

import android.content.Context;
import android.util.SparseIntArray;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;

import com.ashlikun.adapter.ViewHolder;
import com.ashlikun.adapter.recyclerview.vlayout.SingAdapter;

import java.util.List;

/**
 * 作者　　: 李坤
 * 创建时间: 2018/4/10 0010　20:53
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：VLayout的ItemAdapter
 * <p>
 * 多表单的内部adapter,可以有多个不同的type,
 * 第一步在构造方法addItemType关联id
 */
public abstract class MultipleSingAdapter<T> extends SingAdapter<T> {
    /**
     * ItemType对应的LayoutId
     */
    private SparseIntArray layouts;
    public static final int TYPE_NOT_FOUND = -404;
    public static final int DEFAULT_VIEW_TYPE = -0xff;

    public MultipleSingAdapter(Context context, List<T> datas) {
        super(context, datas);
    }

    @Override
    public int getItemViewType(int position) {
        return getItemViewType(position, getItemData(position));
    }


    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        int layoutId = getLayoutId(viewType);
        ViewHolder holder = new ViewHolder(context, createRoot(parent, layoutId, viewType), this);
        setListener(parent, holder, viewType);
        return holder;
    }

    /**
     * 关联 viewType与layoutResId
     *
     * @param viewType
     * @param layoutResId
     */
    public void addItemType(int viewType, @LayoutRes int layoutResId) {
        if (layouts == null) {
            layouts = new SparseIntArray();
        }
        layouts.put(viewType, layoutResId);
    }

    /**
     * 添加默认的view
     *
     * @param layoutResId
     */
    protected void setDefaultViewTypeLayout(@LayoutRes int layoutResId) {
        addItemType(DEFAULT_VIEW_TYPE, layoutResId);
    }

    @Override
    public int getLayoutId(int viewType) {
        return layouts.get(viewType, TYPE_NOT_FOUND);
    }

    public abstract int getItemViewType(int position, T data);
}

File: adapter/src/main/java/com/ashlikun/adapter/recyclerview/multiltem/MultipleSingAdapter.kt
package com.ashlikun.adapter.recyclerview.multiltem

import android.content.Context
import com.ashlikun.adapter.recyclerview.vlayout.SingAdapter
import androidx.annotation.LayoutRes
import androidx.viewbinding.ViewBinding
import com.alibaba.android.vlayout.LayoutHelper
import com.alibaba.android.vlayout.layout.LinearLayoutHelper
import com.ashlikun.adapter.recyclerview.AdapterConvert
import com.ashlikun.adapter.recyclerview.vlayout.mode.AdapterBus
import com.ashlikun.adapter.recyclerview.vlayout.mode.LayoutStyle

/**
 * @author　　: 李坤
 * 创建时间: 2021/12/5 1:10
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：VLayout的ItemAdapter
 * 多表单的内部adapter,可以有多个不同的type,
 * 第一步在构造方法addItemType关联id
 */

open class MultipleSingAdapter<T>(
    context: Context,
    initDatas: List<T>? = null,
    //事件
    override var bus: AdapterBus? = null,
    //data对应的type
    open var itemType: ((position: Int, data: T) -> Int)? = null,
    //布局，优先
    layoutStyle: LayoutStyle? = null,
    //ViewType
    override var viewType: Any? = null,
    //初始化的apply 便于执行其他代码
    apply: (MultipleSingAdapter<T>.() -> Unit)? = null,
    //转换
    override var convert: AdapterConvert<T>? = null
) : SingAdapter<T>(
    context = context,
    initDatas = initDatas,
    layoutStyle = layoutStyle
) {
    //ItemType对应的binding 多个
    open var bindings: MutableMap<Int, Class<out ViewBinding>> = mutableMapOf()

    //ItemType对应的LayoutId
    open var layouts: MutableMap<Int, Int> = mutableMapOf()

    init {
        apply?.invoke(this)
    }

    override fun getItemViewType(position: Int): Int {
        var data: T = getItemData(position) ?: return super.getItemViewType(position)
        return getItemViewType(position, data)
    }

    /** 关联 viewType与layoutResId
     *
     * @param viewType
     * @param layoutResId
     */
    fun addItemType(viewType: Int, @LayoutRes layoutResId: Int) {
        layouts[viewType] = layoutResId
    }

    /**
     * 关联 viewType与ViewBinding
     *
     * @param viewType
     * @param layoutResId
     */
    fun addItemType(viewType: Int, viewBindClass: Class<out ViewBinding>) {
        bindings[viewType] = viewBindClass
    }

    override fun getLayoutId(viewType: Int): Int {
        return layouts[viewType] ?: TYPE_NOT_FOUND
    }

    override fun getBindClass(viewType: Int): Class<out ViewBinding>? {
        return bindings[viewType]
    }

    /**
     * 添加默认的view
     *
     * @param layoutResId
     */
    protected fun setDefaultViewTypeLayout(@LayoutRes layoutResId: Int) {
        addItemType(DEFAULT_VIEW_TYPE, layoutResId)
    }

    /**
     * 添加默认的view
     *
     * @param layoutResId
     */
    protected fun setDefaultViewTypeLayout(viewBindClass: Class<out ViewBinding>) {
        addItemType(DEFAULT_VIEW_TYPE, viewBindClass)
    }

    open fun getItemViewType(position: Int, data: T) =
        itemType?.invoke(position, data) ?: throw RuntimeException("必须提供itemtype")

    companion object {
        const val TYPE_NOT_FOUND = -404
        const val DEFAULT_VIEW_TYPE = -0xff
    }
}